### PR TITLE
fix(codegen): dispatch dual_aiv_dispatch AIV kernels on both vector lanes

### LINKED ISSUE #2006: [Bug] pl.split_aiv(2, SplitMode.NONE) silently drops half the work — one AIV lane per block never writes back
## Background

`pl.split_aiv(2, mode=pl.SplitMode.NONE)` **silently drops half the work**. Exactly one of the two
AIV lanes per SPMD block has its `pl.assemble` writes discarded — the destination rows come back as
**zeros, never written**. It compiles clean, raises nothing, produces no NaN.

Found while debugging qwen3-14b's fused decode (`models/qwen3/14b/decode_layer.py`), whose rope and
online-softmax phases both run inside `pl.split_aiv(2, SplitMode.NONE)` regions. The kernel returns
wrong tokens, and the in-repo golden cannot see it because that golden is a second on-device dispatch
of the same code path.

Reproduction environment:

| Component | Version |
|---|---|
| pypto | `8ebddcb8` (`main`) — also reproduces on `a655a5a7` |
| simpler | `c94aa9f3` |
| ptoas | `0.48` |
| pto-isa | `83d01313` (the pin) |
| CANN | `9.0.0` |
| NPU | Ascend 910B (`a2a3`) |

Independent of the two other pypto defects found in the same investigation — the orchestration
forwarded-tuple mis-bind (#2003 / PR #2004) and the `tile.concat` missing `.not_inplace_safe()`.
This reproduces with both of those fixed.

### Component

Codegen (`LowerAutoVectorSplit`, pass 20)

## The documented contract

From `pl.split_aiv`'s own docstring (`python/pypto/language/dsl_api.py:850`):

> * **Task-parallel** (`NONE`): **no halving** — both lanes run the *full* body for disjoint work
>   the author dispatches via `aiv_id` (e.g. an `aiv_id`-strided loop). Use this when the tiles
>   cannot be halved (unit dims) or a reduction must stay full-width.

The reproducer does exactly that: 16 SPMD blocks × 2 AIV lanes = 32 lanes, each lane owning 4
disjoint items dispatched via `aiv_id`. No halving is expected or wanted.

## Actual behaviour

```
CONTROL  plain pl.spmd(32), no split_aiv       PASS   max_err= 0.00000   wrong items =   0/128
BROKEN   pl.spmd(16) + pl.split_aiv(2, NONE)   FAIL   max_err= 4.18413   wrong items =  64/128
    wrong lanes (16/32): [1, 2, 5, 6, 9, 10, 13, 14, 17, 18, 21, 22, 25, 26, 29, 30]
    item 4 (lane 1) actually holds: nothing recognisable
    want[4][0,:4] = [-1.0464614629745483, -0.6825683116912842, 0.4256860017776489, 0.5112801790237427]
    got [4][0,:4] = [0.0, 0.0, 0.0, 0.0]
```

The dropped items are **exactly zero** — they were never written, not miscomputed.

**The pattern is perfectly regular.** With `lane = pl.get_block_idx() * 2 + aiv_id`:

| SPMD block | lane (aiv=0) | lane (aiv=1) |
|---|---|---|
| 0 (even) | ✅ written | ❌ **dropped** |
| 1 (odd)  | ❌ **dropped** | ✅ written |
| 2 (even) | ✅ written | ❌ **dropped** |
| 3 (odd)  | ❌ **dropped** | ✅ written |

i.e. **exactly one lane per block survives, and which one alternates with the block's parity.**

## Steps to reproduce

Standalone; imports only `pypto.language` + `pypto.runtime`. No pypto-lib, no golden harness.

```bash
python pypto_split_aiv_bug_repro.py -d <device_id>     # exits 1 on the bug
```

<details>
<summary><b>pypto_split_aiv_bug_repro.py</b></summary>

```python
import argparse

import torch

import pypto.language as pl
from pypto.runtime import RunConfig

LANES = 32           # AIV lanes
IPL = 4              # items per lane
ITEMS = LANES * IPL  # 128
ROWS = 5
COLS = 128
SRC_ROWS = ITEMS
OUT_ROWS = ITEMS * ROWS


@pl.jit
def split_aiv_arm(
    src: pl.Tensor[[SRC_ROWS, ROWS * COLS], pl.FP32],
    out: pl.Out[pl.Tensor[[OUT_ROWS, COLS], pl.FP32]],
):
    """16 spmd blocks x 2 AIV lanes = 32 lanes. Each lane owns 4 disjoint items."""
    with pl.spmd(16, name_hint="sa"):
        for aiv in pl.split_aiv(2, mode=pl.SplitMode.NONE):
            lane = pl.get_block_idx() * 2 + aiv
            for it in pl.range(IPL):
                g = lane * IPL + it
                out = pl.assemble(out, pl.reshape(src[g : g + 1, 0 : ROWS * COLS], [ROWS, COLS]),
                                  [g * ROWS, 0])
    return out


@pl.jit
def plain_spmd_arm(
    src: pl.Tensor[[SRC_ROWS, ROWS * COLS], pl.FP32],
    out: pl.Out[pl.Tensor[[OUT_ROWS, COLS], pl.FP32]],
):
    """CONTROL: identical work partition, 32 spmd blocks, no split_aiv."""
    with pl.spmd(LANES, name_hint="ps"):
        lane = pl.get_block_idx()
        for it in pl.range(IPL):
            g = lane * IPL + it
            out = pl.assemble(out, pl.reshape(src[g : g + 1, 0 : ROWS * COLS], [ROWS, COLS]),
                              [g * ROWS, 0])
    return out


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("-p", "--platform", default="a2a3")
    ap.add_argument("-d", "--device", type=int, default=0)
    args = ap.parse_args()

    torch.manual_seed(1)
    src = torch.randn(SRC_ROWS, ROWS * COLS, dtype=torch.float32)
    want = src.reshape(OUT_ROWS, COLS)
    cfg = RunConfig(platform=args.platform, device_id=args.device)

    failed = []
    for tag, fn in (("CONTROL  plain pl.spmd(32), no split_aiv", plain_spmd_arm),
                    ("BROKEN   pl.spmd(16) + pl.split_aiv(2, NONE)", split_aiv_arm)):
        out = torch.zeros(OUT_ROWS, COLS, dtype=torch.float32)
        fn(src, out, config=cfg)
        o = out.view(ITEMS, ROWS, COLS)
        w = want.view(ITEMS, ROWS, COLS)
        bad = [g for g in range(ITEMS) if not torch.equal(o[g], w[g])]
        err = float((out - want).abs().max())
        print(f"[split_aiv] {tag:<46} {'PASS' if not bad else 'FAIL'}   "
              f"max_err={err:8.5f}   wrong items = {len(bad):3d}/{ITEMS}", flush=True)
        if bad:
            failed.append(tag)
            lanes_bad = sorted({g // IPL for g in bad})
            print(f"[split_aiv]     wrong lanes ({len(lanes_bad)}/{LANES}): {lanes_bad}")
            g0 = bad[0]
            print(f"[split_aiv]     want[{g0}][0,:4] = {w[g0][0, :4].tolist()}")
            print(f"[split_aiv]     got [{g0}][0,:4] = {o[g0][0, :4].tolist()}")

    print(f"\n[split_aiv] {'BUG REPRODUCED' if failed else 'no bug on this build'}", flush=True)
    raise SystemExit(1 if failed else 0)
```

</details>

The two arms have the **identical work partition**, identical inputs, identical golden. The only
difference is `pl.split_aiv(2, NONE)` vs a plain `pl.spmd(32)`.

## Ablation — `pl.split_aiv` is the only necessary ingredient

Bisected from qwen3's real phase-0 body (an in-register QK-norm) down to the repro above. Each row
removes exactly ONE ingredient; work partition, inputs and golden are identical throughout.

| variant | `split_aiv`? | result |
|---|:---:|---|
| full chain, `spmd + split_aiv + lane-guard + pl.pipeline` | ✅ | **FAIL** 64/128 |
| `pl.range` instead of `pl.pipeline` | ✅ | **FAIL** 64/128 |
| lane guard removed | ✅ | **FAIL** 64/128 |
| **`split_aiv` removed (plain `pl.spmd(32)`)** | ❌ | **PASS** 0/128 |
| all QK-norm arithmetic removed (`concat`+`reshape` only) | ✅ | **FAIL** 64/128 |
| **`pl.concat` removed too (slice + reshape only)** | ✅ | **FAIL** 64/128 |
| `pl.full` hoisted out of the loop | ✅ | **FAIL** 64/128 |

So it needs **no** `pl.concat`, **no** reduction, **no** arithmetic, **no** lane guard, and **no**
`pl.pipeline`. Only `pl.split_aiv`.

The same chain in a plain `pl.at(pl.Level.CORE_GROUP)` task is **bit-exact** against torch
(`max_abs_err = 0.000000`, 0/128 items wrong).

## Suspected mechanism (not verified — for the fix's author)

The body updates a loop-carried SSA tensor:

```python
for aiv in pl.split_aiv(2, mode=pl.SplitMode.NONE):
    ...
    out = pl.assemble(out, tile, [row, 0])      # both lanes produce a new `out`
```

Both lanes write, so both produce a new version of `out`. The dropped rows come back as **exact
zeros — never written**, which is what one would expect if only ONE lane's `out` survives into the
enclosing scope and the other lane's SSA chain is discarded. The block-parity alternation of *which*
lane survives suggests the surviving version is selected by something that flips with the block index.

Per the docstring, the region is consumed and erased by **`LowerAutoVectorSplit` (pass 20)** and
"never reaches ExpandMixedKernel or codegen" — so that is where I would look first.

## Blast radius

`pl.split_aiv(2, SplitMode.NONE)` with an `aiv_id`-dispatched `pl.assemble` is exactly the pattern
qwen3-14b's fused decode uses in **both** its rope phase and its online-softmax phase. Measured
consequence, grading the fused kernel against the un-fused one (a different kernel, and the one that
passes the serving e2e) on a fixture where every batch row shares one `seq_len`:

| seq_len | max_abs_err | argmax agreement |
|---|---|---|
| 6 (a real serving prompt) | **1.65** | **11/16** |
| 31 | 0.62 | 15/16 |
| 128 | 0.19 | 16/16 |
| 1024 | 0.03 | 16/16 |

Logit scale is ~7, so at a short prompt the kernel is 23 % of scale off and flips 5 of 16 tokens; the
error decays with context length, so long fixtures hide it entirely. The qwen3-14b **serving e2e
fails**. pypto-lib previously REVERTED the whole fused-decode feature (pypto-lib#703, "Revert Qwen3
decode changes causing incorrect serving output") — this is one of the reasons why.

## Ruled out

Each was tested and falsified, so please don't re-chase them:

- **Not `pl.concat`.** Removing it entirely still fails (ablation row 6). (The separate
  `tile.concat` missing-`.not_inplace_safe()` defect is real and already fixed; this reproduces on
  top of that fix.)
- **Not the orchestration forwarded-tuple mis-bind** (#2003 / PR #2004). Reproduces with that patch applied.
- **Not the lane guard** (`if lane < N`), **not `pl.pipeline`**, **not `pl.full` placement**,
  **not the arithmetic** — every one removed independently, still fails.
- **Not `pl.reshape`** — the same reshape in a plain `pl.spmd` control is bit-exact.
- **Not pto-isa, not the ring arena, not a race** — deterministic across runs.

## Suggested direction

If the "both lanes run the full body for disjoint work" contract is meant to hold, then
`LowerAutoVectorSplit` must merge (not select) the two lanes' writebacks to a carried tensor.

If instead that pattern is **not** supported — i.e. a carried tensor may not be `pl.assemble`d from
both lanes of a `NONE` region — then the DSL should **reject it at parse time** rather than silently
dropping half the work. `pto.alloc_multi_tile` already takes exactly that stance for a comparable
silent-overlap hazard ("...adjacent slots would silently overlap. `pto.alloc_multi_tile` therefore
rejects `compact = row_plus_one` slot layouts"). Silent wrong data is the worst possible outcome, and
in this case it cost a shipped feature a revert.

### PR DESCRIPTION
**Pull request by luohuan19**
## Summary

A `pl.split_aiv` region is task-parallel: both AIV lanes run the body on disjoint work selected by `aiv_id`. `SplitVectorKernel` stamps such an AIV kernel `dual_aiv_dispatch`, but only the mixed-kernel Group submit honoured it. The three single-kernel dispatch paths — direct call, Spmd wrapper, and AIV-only Group — all built their submit through `CoreTypeToSubmitPrefix()`, which unconditionally yields `rt_submit_aiv_task` for VECTOR.

`rt_submit_aiv_task` fills only the AIV0 slot of `MixedKernels`, so the runtime schedules a single-AIV task: one AIV core per block. The second lane never launches, and the lone lane that does reads a `get_sub_block_id()` the runtime leaves undefined for single-AIV tasks, so its `aiv_id` is really its physical cluster position. **Half the work is silently dropped** — the destination rows come back as exact zeros, with no error, no NaN, and a block-parity pattern for which lane survives. A pure-vector kernel (no cube work) hits this on the Spmd-wrapper path.

### Changes

1. **`fix(language)`: make `split_aiv` IR survive print → parse unchanged.** Two parser-side asymmetries made `python_print` → parse lossy for a function carrying a `pl.split_aiv` region, failing the roundtrip verifier on every pass over such IR:
   - An explicit `attrs={"split": pl.SplitMode.NONE}` was dropped on parse while the printer emits it. Store every mode instead — behaviour-preserving, since `Function::GetSplitMode()` maps `split == 0` to `nullopt` and is the attr's sole reader.
   - A `pl.split_aiv` loop directly in an InCore function body was wrapped in a synthesized `InCoreScopeStmt`. An InCore function is already a core function — nothing is left to outline — and it is the shape post-outline IR prints as. Emit the region in place there too.

2. **`fix(codegen)`: dispatch `dual_aiv_dispatch` AIV kernels on both vector lanes.** Route the three single-kernel paths through a shared `BuildSingleKernelSubmitExpr()`, which emits `MixedKernels{INVALID_KERNEL_ID, id, id}` + `rt_submit_task` for a `dual_aiv_dispatch` VECTOR kernel. Two active AIV slots make it a MIX-shape task, so the scheduler places both lanes of one cluster under the same `block_idx` and gives them `sub_block_id` 0 and 1; the cluster's AIC core is unused. **No runtime change is needed.**

A plain vector kernel carries no `dual_aiv_dispatch` and keeps `rt_submit_aiv_task`, so it retains its full independent-AIV parallelism. Mixed kernels are untouched.

## Testing

- [x] New roundtrip tests (`tests/ut/ir/printing/test_split_aiv_roundtrip.py`) guard both parser arms.
- [x] New orchestration codegen UT (`tests/ut/codegen/test_orchestration_codegen_more.py`) covers the `dual_aiv_dispatch` submit shape.
- [x] The existing NONE-region codegen ST compiled this exact shape but only asserted a `.pto` was emitted — which is why the defect shipped. It now also asserts the submit launches both lanes.
- [x] Documentation updated (`docs/en/dev/codegen/01-orchestration_codegen.md` + zh-cn).

## Related Issues

Fixes #2006

### CODE REVIEW

File: python/pypto/language/parser/ast_parser.py [hunk 1]
```diff
@@ -4146,7 +4146,14 @@ def _parse_split_aiv_for_loop(self, stmt: ast.For, iter_call: ast.Call) -> None:
...
         # enclosing core function and the nested region survives.
-        if self._is_inside_scope(ir.ScopeKind.InCore):
+        #
+        # An ``@pl.function(type=pl.FunctionType.InCore)`` body is already a core
+        # function, so the region goes in place there too: there is nothing left for
+        # OutlineIncoreScopes to outline. This is the shape post-outline IR prints as
+        # (outlining consumes the InCoreScopeStmt and leaves the region directly in an
+        # InCore function body), so synthesizing a wrapper here would make print->parse
+        # lossy — the reparsed function would grow a spurious InCoreScopeStmt.
+        if self._is_inside_scope(ir.ScopeKind.InCore) or self._func_type == ir.FunctionType.InCore:
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/language/parser/ast_parser.py]: Good catch — confirmed and fixed in `e7fd8298`.

I reproduced it: an `@pl.function(type=pl.FunctionType.AIV)` body with a `pl.split_aiv` region did grow a spurious `InCoreScopeStmt` under the `InCore`-only guard.

Rather than an inline tuple, I extracted the set as `_CORE_FUNCTION_TYPES = {InCore, AIC, AIV}`, which mirrors the existing C++ predicate `IsInCoreType()` (`include/pypto/ir/function.h:291`) — that is the canonical definition of "is a core function" in the codebase, and it matches your suggested set exactly.

Guarded by a new roundtrip test (`test_roundtrip_aiv_function_body`) asserting no wrapper scope is synthesized and that the function survives print -> parse unchanged.

**Comment by luohuan19** [comment]
Superseded by https://github.com/hw-native-sys/pypto/pull/2379 — closing this one.

This branch was opened against a month-old `main` and no longer applies (the merge
state has been `CONFLICTING` since). Rather than force-push a rebase over the review
history here, the rebased work is a fresh PR against `main` @ `289290e6`.

What changed while rebasing:

- The codegen fix itself carries over unchanged in substance, re-expressed against
  the refactored `TaskDispatchPlan` / `EffectiveLaunchSpec` API.
- The `fix(language): make split_aiv IR survive print -> parse unchanged` commit is
  dropped. https://github.com/hw-native-sys/pypto/pull/2298 has since made *absence*
  the canonical encoding of `SplitMode::NONE` at the function-attr level, solving the
  same problem the opposite way.
- The parser change from the review-response commit (treating a core function body as
  an implicit InCore scope so a bare `pl.split_aiv` is emitted in place) is dropped and
  left for its own PR: `main` now pins the opposite behaviour in
  `test_scope_nested_region_rejected` / `test_scope_nested_region_guards_not_bypassed`,
  which require a top-level region in an InCore body to be wrapped and then rejected by
  `LowerAutoVectorSplit`. That is a semantics decision, not a rebase artefact.

The defect itself is still live on `main` — re-verified on an Ascend 910B, 16 of 32
lanes never written, with exactly the lane set reported in
https://github.com/hw-native-sys/pypto/issues/2006. One caveat worth recording for
anyone re-checking: the reproducer in that issue **self-masks**. Its plain-`pl.spmd`
control arm runs first and leaves its (correct) rows in the reused output device
buffer, so the dropped lanes read back correct and both arms report PASS. Run the
`split_aiv` arm first and alone to see the failure.
